src/lib.rs

```diff
@@ -24,6 +24,17 @@ impl NavigationManager {
     }
 }
 
+macro_rules! with_navigator {
```

Comment:
It's possible to short even more, something like:

```
macro_rules! with_navigator {
    ($func:ident ( $( $arg:expr ),* )) => {
        NavigationManager::get_instance()
            .lock()
            .unwrap()
            .as_mut()
            .unwrap()
            .navigator
            .$func( $( $arg ),* )
    }
}
```



If you use the macro like this, you can just:
       
```
        fn self_test() -> bool {
            with_navigator!(self_test())
        }

        fn set_pwm_channel_value(channel: PwmChannel, value: u16) {
            with_navigator!(set_pwm_channel_value(channel.into(), value))
        }
```

Reply:
My original point is it imitate a function call, and not a function that received a lambda, since this will break style for big functions or builder pattters, like:
```rs
with_navigator!()
    .set_pwm_channel_value(
        array_channels[i].clone().into(),
        array_values[i]
     );

// vs
with_navigator!(
    set_pwm_channel_value(
        array_channels[i].clone().into(),
        array_values[i]
    )
);
```
and
```rs
let k = with_navigator!()
    .a()
    .b()
    .c();
    
//vs
let k = with_navigator!(
    .a()
    .b()
    .c()
);
```
But we can get also feedback from @joaoantoniocardoso

Reply:
Now comparing.. I also don't see it as more compact..

Reply:
Ok, it's good to keep the cargo fmt

Reply:
I prefer `with_navigator!().a().b().c()` syntax as it better resembles the original code, and doesn't hide the fact that the mutex will still be in scope after the call.

Another option that would provide a closed scope would be something like:
```Rust
let k = with_manager(|manager| manager.navigator.a().b().c() );
```

which doesn't need to be a macro at all:

```Rust
fn with_mut_manager<F, R>(closure: F) -> R
where
    F: FnOnce(&mut NavigationManager) -> R,
{
    let mut guard = NavigationManager::get_instance()
        .lock()
        .unwrap();

    let mut navigator = guard
        .as_mut()
        .unwrap();

    closure(&mut navigator)
}
```

Reply:
it's elegant,

just to register we can also use the closure like follows:

```
use navigator_rs::Navigator
let reading = NavigatorManager::with_mut( Navigator::read_all());
```

vs

actual
```
with_navigator!().read_all()
```


It seems a bit more complex to write, but you have a verbose function calling NavigatorManager to handle the action.

Reply:
What are the benefits for this approach ?

Reply:
seems more verbose, you can visualise that NavigatorManager holds the Navigator object's usage.

Reply:
Verbose is not a benefit..

Reply:
The closed scope for the manager guard is the real benefit here, which for maintainability means a less error-prone mutex interface, especially because by using functions to access the manager, we can separate the static MANAGER into another mod, enforcing the use of this API, as shown in the example below:

```Rust

pub mod manager {
    use std::sync::{Arc, Mutex};

    lazy_static! {
        static ref MANAGER: Arc<Mutex<Manager>> = Default::default();
    }

    #[derive(Default)]
    pub struct Manager {
        pub data: u8,
    }

    pub fn with_ref<F, R>(closure: F) -> R
    where
        F: FnOnce(&Manager) -> R,
    {
        let manager = MANAGER.lock().unwrap();

        closure(&manager)
    }

    pub fn with_mut<F, R>(closure: F) -> R
    where
        F: FnOnce(&mut Manager) -> R,
    {
        let mut manager = MANAGER.lock().unwrap();

        closure(&mut manager)
    }
}

pub fn increment_data() -> u8 {
    manager::with_mut(|m| {
        m.data += 1;
        m.data
    })
}

pub fn read_data() -> u8 {
    manager::with_ref(|m| m.data)
}

```